File: app/src/main/java/home/self/beerviewer_mvvm/view/beersview/BeersViewModel.java
package home.self.beerviewer_mvvm.view.beersview;

import androidx.databinding.BaseObservable;

import com.orangegangsters.github.swipyrefreshlayout.library.SwipyRefreshLayoutDirection;

import javax.inject.Inject;

import home.self.beerviewer_mvvm.data.source.BeerRepository;
import home.self.beerviewer_mvvm.rx.rxbus.Events;
import home.self.beerviewer_mvvm.rx.rxbus.RxEventBus;
import home.self.beerviewer_mvvm.rx.schedulers.BaseSchedulerProvider;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class BeersViewModel extends BaseObservable {

    private BeerRepository beerRepository;
    private BaseSchedulerProvider schedulerProvider;

    private BeersViewNavigator beersView;

    private CompositeDisposable compositeDisposable;

    @Inject
    public BeersViewModel(BeerRepository beerRepository, BaseSchedulerProvider schedulerProvider) {
        this.beerRepository = beerRepository;
        this.schedulerProvider = schedulerProvider;

        this.compositeDisposable = new CompositeDisposable();
        onEventBusCalled();
    }

    void getBeers(int pageStart, int perPage, SwipyRefreshLayoutDirection direction) {
        compositeDisposable.clear();
        Disposable disposable = beerRepository
                .getBeers(pageStart, perPage)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(beers-> {
                    if (direction == SwipyRefreshLayoutDirection.TOP)
                        beersView.showItems(beers);
                    else if (direction == SwipyRefreshLayoutDirection.BOTTOM)
                        beersView.showItemsFromBottom(beers);
                });

        compositeDisposable.add(disposable);
    }

    private void onEventBusCalled() {
        Disposable disposable = RxEventBus.getInstance().getBusObservable()
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(event -> {
                    if (event instanceof Events.PageEvent)
                        beersView.setPageStart();
                });

        compositeDisposable.add(disposable);
    }

    public void takeView(BeersViewNavigator beersView) {
        this.beersView = beersView;
    }
}

File: app-kotlin/src/main/java/home/self/beerviewer_mvvm/app_kotlin/view/beersview/BeersViewModel.kt
package home.self.beerviewer_mvvm.app_kotlin.view.beersview

import androidx.lifecycle.MutableLiveData
import com.orangegangsters.github.swipyrefreshlayout.library.SwipyRefreshLayoutDirection
import home.self.beerviewer_mvvm.app_kotlin.BaseViewModel
import home.self.beerviewer_mvvm.app_kotlin.Constants
import home.self.beerviewer_mvvm.app_kotlin.data.model.BeerModel
import home.self.beerviewer_mvvm.app_kotlin.data.source.BeerRepositoryApi
import home.self.beerviewer_mvvm.app_kotlin.di.qualifier.App
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject
import javax.inject.Inject

internal interface BeersViewModel {

    interface Inputs {
        fun fetchBeers(pageStart: Int, perPage: Int, direction: SwipyRefreshLayoutDirection): Disposable
    }

    interface Outputs {
        fun fetchBeers(): Observable<List<BeerModel>>
        fun fetchIndex(): Observable<Int>
        fun isLoading(): Observable<Boolean>
    }

    class ViewModel @Inject constructor(
            @App private val repository: BeerRepositoryApi

    ): BaseViewModel(), Inputs, Outputs {

        val intpus = this
        val outpus = this

        private val beers = BehaviorSubject.create<List<BeerModel>>()
        private val isLoading = PublishSubject.create<Boolean>()
        private val index = PublishSubject.create<Int>()

        init {
            intent().map { it.getSerializableExtra(Constants.EXTRA_DEFAULT_PAGE) as Pair<Int, Int> }
                    .doOnNext { isLoading.onNext(true) }
                    .subscribeBy { it ->
                        fetchBeers(it.first, it.second, SwipyRefreshLayoutDirection.TOP)
                    }

            subscribeIndex()
        }

        override fun fetchBeers(pageStart: Int, perPage: Int, direction: SwipyRefreshLayoutDirection): Disposable
                = repository
                .fetchBeers(pageStart, perPage)
                .doOnNext { isLoading.onNext(false) }
                .flatMap { beers ->
                    Flowable.fromIterable(beers)
                            .doOnNext { beer -> beer.direction = direction }
                            .toList()
                            .toFlowable()
                }
                .subscribeBy { beerList ->
                    beers.onNext(beerList)
                }

        override fun fetchBeers(): Observable<List<BeerModel>> = beers

        override fun fetchIndex(): Observable<Int> = index

        override fun isLoading(): Observable<Boolean> = isLoading

        private fun subscribeIndex()
                = repository
                .getIndex()
                .subscribe { it ->
                    index.onNext(it)
                }
    }

}

